text/0000-parallel-model-hooks.md

```diff
@@ -0,0 +1,115 @@
+This has two major implications:
+
+1. The `afterModel` hook will be passed a promise, since the result of the `model` hook will not have resolved by the time `afterModel` is called.
+2. Routes may not have resolved the data requested by `modelFor` when it is called. It will need to return a promise or there will need to be another method for getting a promise that eventually resolves a route's model. (`modelPromiseFor`?)
+
+#### Router.js (tildeio/router.js#171)
+
+`handler-info` presently has a `resolve` method that calls the model hooks. We move the model hook calls to a new `request` method, and save the promises on the `handler-info`'s context (the promise from the `model` hook is passed to the `afterModel` hook). The `resolve` method takes the stored promises and inserts them into the resolution chain.
+
+`transition-state` presently has a `resolveOneHandlerInfo` method that recurses through the `handler-info`s for a transition and calls the associated `resolve` method. Before this, we insert a `requestOneHandlerInfo` method that recurses through `handler-info`s, calling each one's `request` method, which now executes the model hooks, kicking off any network requests.
+
+We add a `requestIndex` property to `transition` that keeps track of how many route's model hooks we have already run (akin to how `resolveIndex` presently keeps track of how many routes we have resolved).
+
+#### Ember.js (emberjs/ember.js#12415)
+
+The default model hook uses `requestIndex` (instead of `resolveIndex`) to grab `modelPromise` (instead of `context`) from the `handler-info`.
+
+`Route#modelFor` returns a promise.
+
+#### Usage
+
+Routes that don't depend on another route's data likely don't need any changes, but their requests are sent in parallel:
+
+```javascript
+// app/routes/parent.js
+export default Ember.Route.extend({
+  model() {
+    return Ember.$.get('/api/foo');
+  }
+});
+
+// app/routes/parent/child.js
+export default Ember.Route.extend({
+  model(params) {
+    return Ember.$.get(`/api/foo/${params.id}`);
+  }
+});
+```
+
+Routes that depend on another route's data must consume that data as a promise:
+
+```javascript
+// app/routes/parent.js
+export default Ember.Route.extend({
+  model(params) {
+    return Ember.$.get(`/api/foo/${params.id}`);
+  }
+});
+
+// app/routes/parent/child.js
+export default Ember.Route.extend({
+  model() {
+    return this.modelFor('parent').then((parentModel) => {
+      return Ember.$.get(`/api/bar/${parentModel.get('baz')}`);
+    });
+  }
+});
+```
+
+Routes whose `afterModel` hook operates on the resolved model must now consume a promise:
+
+```javascript
+export default Ember.Route.extend({
+  model(params) {
+    return Ember.$.get(`/api/foo/${params.id}`);
+  },
+  afterModel(modelPromise, transition) {
+    return modelPromise.then((data) => {
+      if (data.abort) {
+        this.transitionTo('fiz');
+      }
+    }
+  }
+});
+```
+
+# Drawbacks
+
+- This is a major breaking change.
```

Comment:
To help explore the problem space I'd love to get an early version in behind a feature flag so we have fewer unknowns. It seems like it would be a good fit to come up with an API for this sort of behavior at the same time as trying to solidify routable components so that they play nicely together. It also gives us a `GlimmerComponent`/`"use strict";` type of escape valve for opting in if we can't come up with a better way to handle it.

The option of building an addon has community effects that will be quite painful. We see tight coupling between liquid-fire and Ember that an addon to accomplish this would make look positively decoupled. Self-trolling will occur all over the place.